### ISSUE TITLE
Impossible to select a void inline in Safari 11.1

### ISSUE THREAD

**Issue by justinweiss** [ianstormtaylor/slate#1852]
#### Do you want to request a _feature_ or report a _bug_?

Report a bug.

#### What's the current behavior?

In Safari 11.1, you can't select a void inline by clicking on it. You also can't navigate through it from the left side with the arrow keys.

I'm pretty sure this used to work, but I can't tell whether it was a bug introduced by Safari, or if it was a result of the change to empty text nodes on void nodes, since I made both changes in my demo app around the same time.

It appears that Safari will always force the selection to the end of the preceding text node, which causes this problem. I wonder if it's not considering zero-width spaces when deciding on the size of the inline.

Works fine in Chrome / Firefox.

Fiddle here (open your console, and it will print the list of selected inlines on every change): https://jsfiddle.net/oyerecen/

#### What's the expected behavior?

Clicking on the inline void node should select the node, and you should be able to use the keyboard to navigate through it.

**Comment by justinweiss** [comment]
A few things I’ve seen so far:

When positioning a cursor at offset 0 of a node, Safari will normalize the native selection to be the end of the previous node. Chrome will allow offset 0 as a valid position. 

It seems that during the process of updating the selection, onNativeSelectionChange is called twice. I think this is because of the throttle, that it’s called on the trailing edge of the 100ms. This will almost certainly happen too late to be blocked by tmp.isSelectionUpdating.

So I think this is what’s happening. We have a native selection update on offset 0 of the inline node. This triggers onNativeSelectionChange, which triggers the onSelect stack. That sets the native selection to offset 0, which gets normalized to offset `previousNode.length`. This triggers another onNativeSelectionChange, which gets throttled but still happens later. This will go back through the plugin stack, setting the Slate selection to offset `previousNode.length`, which registers as outside of the inline. 

I don’t know for sure that this is what’s happening, but it seems to match what I’m seeing. I’ll dig into it more.

**Comment by ianstormtaylor** [comment]
I believe that this may be fixed by https://github.com/ianstormtaylor/slate/pull/3093, which has changed a *lot* of the logic in Slate and `slate-react` especially. I'm going to close this out, but as always, feel free to open a new issue if it persists for you. Thanks for understanding.
